File: C++/other/binary_trie.cpp
struct binary_trie{
    struct node{
        int ch[2];
        int count;
        node() : ch{-1,-1}, count(0) {}
    };

    static constexpr int LOG = 60;
    vector<node> T;

    binary_trie(){
        T.emplace_back();
    }

    int new_node(){
        T.emplace_back();
        return (int)T.size()-1;
    }

    int size() const{
        return T[0].count;
    }

    int count(ll x) const{
        int v = 0;
        for (int i = LOG-1; i >= 0; i--){
            int f = (x>>i)&1;
            if (T[v].ch[f] == -1) return 0;
            else v = T[v].ch[f];
        }
        return T[v].count;
    }

    void insert(ll x, int k = 1){
        int v = 0;
        T[v].count += k;
        for (int i = LOG-1; i >= 0; i--){
            int f = (x>>i)&1;
            if (T[v].ch[f] == -1) T[v].ch[f] = new_node();
            v = T[v].ch[f];
            T[v].count += k;
        }
    }

    bool erase(ll x, int k = 1){
        if (count(x) < k) return false;
        else{
            int v = 0;
            T[v].count -= k;
            for (int i = LOG-1; i >= 0; i--){
                int f = (x>>i)&1;
                v = T[v].ch[f];
                T[v].count -= k;
            }
            return true;
        }
    }

    ll kth(int k) const{
        assert(0 <= k && k < size());
        int v = 0;
        ll x = 0;
        for (int i = LOG-1; i >= 0; i--){
            int lc = 0;
            if (T[v].ch[0] != -1) lc = T[T[v].ch[0]].count;
            if (k < lc) v = T[v].ch[0];
            else{
                k -= lc;
                x |= 1LL<<i;
                v = T[v].ch[1];
            }
        }
        return x;
    }

    int count_less(ll x) const{
        int v = 0;
        int res = 0;
        for (int i = LOG-1; i >= 0; i--){
            if (v == -1) break;
            int f = (x>>i)&1;
            if (f){
                int u = T[v].ch[0];
                if (u != -1) res += T[u].count;
            }
            v = T[v].ch[f];
        }
        return res;
    }

    ll min_xor(ll x) const{
        assert(size() > 0);
        int v = 0;
        ll res = 0;
        for (int i = LOG-1; i >= 0; i--){
            int f = (x>>i)&1;
            int u = T[v].ch[f];
            if (u != -1 && T[u].count > 0){
                v = u;
            }
            else{
                res |= 1LL<<i;
                v = T[v].ch[f^1];
            }
        }
        return res;
    }

    ll max_xor(ll x) const{
        assert(size() > 0);
        int v = 0;
        ll res = 0;
        for (int i = LOG-1; i >= 0; i--){
            int f = ((x>>i)&1)^1;
            int u = T[v].ch[f];
            if (u != -1 && T[u].count > 0){
                res |= 1LL<<i;
                v = u;
            }
            else v = T[v].ch[f^1];
        }
        return res;
    }
};

File: py/binary_trie.py
class binary_trie:
    class Node:
        __slots__ = ("one", "zero", "count")
        def __init__(self):

            self.one = -1
            self.zero = -1
            self.count = 0

    def __init__(self, log=60):
        self.LOG = log
        self.T = [self.Node()]

    def _new_node(self):
        self.T.append(self.Node())
        return len(self.T)-1

    def size(self):
        return self.T[0].count

    def count(self, x):
        v = 0
        for i in range(self.LOG-1, -1, -1):
            f = (x>>i)&1
            nxt = self.T[v].one if f else self.T[v].zero
            if nxt == -1:
                return 0
            v = nxt
        return self.T[v].count

    def insert(self, x, k=1):
        v = 0
        self.T[v].count += k
        for i in range(self.LOG-1, -1, -1):
            f = (x>>i)&1
            if f:
                if self.T[v].one == -1:
                    self.T[v].one = self._new_node()
                v = self.T[v].one
            else:
                if self.T[v].zero == -1:
                    self.T[v].zero = self._new_node()
                v = self.T[v].zero
            self.T[v].count += k

    def erase(self, x, k=1):
        if self.count(x) < k:
            return False
        v = 0
        self.T[v].count -= k
        for i in range(self.LOG-1, -1, -1):
            f = (x>>i)&1
            v = self.T[v].one if f else self.T[v].zero
            self.T[v].count -= k
        return True

    def kth(self, k):
        assert 0 <= k < self.size()
        v = 0
        x = 0
        for i in range(self.LOG-1, -1, -1):
            lc = 0
            if self.T[v].zero != -1:
                lc = self.T[self.T[v].zero].count
            if k < lc:
                v = self.T[v].zero
            else:
                k -= lc
                x |= 1<<i
                v = self.T[v].one
        return x

    def count_less(self, x):
        v = 0
        res = 0
        for i in range(self.LOG-1, -1, -1):
            if v == -1:
                break
            f = (x>>i)&1
            if f:
                u = self.T[v].zero
                if u != -1:
                    res += self.T[u].count
            v = self.T[v].one if f else self.T[v].zero
        return res

    def min_xor(self, x):
        assert self.size() > 0
        v = 0
        res = 0
        for i in range(self.LOG-1, -1, -1):
            f = (x>>i)&1
            u = self.T[v].one if f else self.T[v].zero
            if u != -1 and self.T[u].count > 0:
                v = u
            else:
                res |= 1<<i
                v = self.T[v].zero if f else self.T[v].one
        return res

    def max_xor(self, x):
        assert self.size() > 0
        v = 0
        res = 0
        for i in range(self.LOG-1, -1, -1):
            f = ((x>>i)&1)^1
            u = self.T[v].one if f else self.T[v].zero
            if u != -1 and self.T[u].count > 0:
                res |= 1<<i
                v = u
            else:
                v = self.T[v].zero if f else self.T[v].one
        return res
